### PR TITLE
Try adding performance tests

### PR DESCRIPTION
**Pull request by opr**
This PR adds some performance tests - they will run in addition to (but separately from) the existing E2E tests. Currently, it doesn't run in CI, the setup for that will be done in another PR to keep the scope of this PR focused.

More info is here: pca54o-3eY-p2

- Create functions to track the highest/lowest/average numbers in an array.
- Create tests that measure times to do actions, and inserts those times into an array.
- Add Performance Reporter to print out the longest/shortest/average times for each performance test

### Screenshots
image

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests
### Manual Testing

How to test the changes in this Pull Request:

1. run `npm run test:performance` watch the results come through 🥳 
 
### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Not required

### Changelog

> Add performance E2E tests.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -67,7 +67,8 @@
 		"storybook:deploy": "rimraf ./storybook/dist/* && npm run storybook:build && gh-pages -d ./storybook/dist",
 		"test": "wp-scripts test-unit-js --config tests/js/jest.config.json",
 		"test:debug": "ndb .",
-		"test:e2e": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js",
+		"test:e2e": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js --testPathIgnorePatterns performance",
+		"test:performance": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js -- performance",
 		"test:e2e-dev": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js --puppeteer-interactive",
```
Comment (alexflorisca) [inline package.json]: Think we should add `--testPathIgnorePatterns performance` for this one and `test:e2e-dev-watch` as well

File: tests/e2e/config/performance-reporter.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		const reportFileContents = readFileSync( PERFORMANCE_REPORT_FILENAME )
+			.toString()
+			.split( '\n' )
+			.slice( 0, -1 )
+			.map( ( line ) => JSON.parse( line ) );
+
+		reportFileContents.forEach( ( testReport ) => {
+			// eslint-disable-next-line no-console
+			console.log(
+				chalk.bgGreenBright.underline.bold( testReport.description )
```
Comment (alexflorisca) [inline tests/e2e/config/performance-reporter.js]: The contrast of light green with white text is pretty hard to read. Maybe we can find a darker green?

File: tests/e2e/config/performance-reporter.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			// eslint-disable-next-line no-console
+			console.log( chalk.red( `Longest: ${ testReport.longest }ms` ) );
+			// eslint-disable-next-line no-console
+			console.log(
+				chalk.green( `Shortest: ${ testReport.shortest }ms` )
+			);
+			// eslint-disable-next-line no-console
+			console.log(
+				chalk.yellow(
+					`Average: ${ testReport.average.toFixed( 2 ) }ms`
```
Comment (alexflorisca) [inline tests/e2e/config/performance-reporter.js]: If longest and shortest don't include any decimal places, I'd leave this as `toFixed()`. We're measuring milliseconds here as well, so probably don't need to go to decimal point of a ms.

File: tests/e2e/specs/performance/cart.test.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+			inserterSearch: [],
+			listViewOpen: [],
+		};
+
+		let i = 3;
+
+		// Measuring loading time.
+		while ( i-- ) {
+			await page.reload();
+			await page.waitForSelector( '.wc-block-cart', { timeout: 120000 } );
```
Comment (alexflorisca) [inline tests/e2e/specs/performance/cart.test.js]: This seems like a mahoosive timeout?

File: tests/e2e/specs/performance/cart.test.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+			logPerformanceResult( `Cart block loading: (${ name })`, value );
+		} );
+
+		// To stop warning about no assertions.
+		expect( true ).toBe( true );
+	} );
+
+	it( 'Quantity change', async () => {
+		await shopper.block.goToCart();
+		await page.waitForNetworkIdle( { idleTime: 2000 } );
```
Comment (alexflorisca) [inline tests/e2e/specs/performance/cart.test.js]: do we need to wait for networkIdle if we're waiting for a selector next?

File: tests/e2e/specs/performance/cart.test.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		}
+		logPerformanceResult(
+			'Cart block: Change cart item quantity',
+			timesForResponse
+		);
+	} );
+
+	it( 'Coupon entry', async () => {
+		await shopper.block.goToCart();
+		await page.waitForNetworkIdle( { idleTime: 2000 } );
```
Comment (alexflorisca) [inline tests/e2e/specs/performance/cart.test.js]: again, might not need this - see comment above

File: tests/e2e/specs/shopper/cart-update-product-quantity.test.js [hunk 7]
```diff
@@ -21,12 +21,12 @@ describe( 'Shopper → Cart → Can update product quantity', () => {
 		await shopper.block.emptyCart();
 	} );
 
-	it( 'allows customer to update product quantity via the input field', async () => {
+	it.only( 'allows customer to update product quantity via the input field', async () => {
```
Comment (alexflorisca) [inline tests/e2e/specs/shopper/cart-update-product-quantity.test.js]: You forgot to remove `.only` here

File: tests/utils/performance.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+export async function getLoadingDurations() {
```
Comment (alexflorisca) [inline tests/utils/performance.ts]: Like github says, this is a ts file so you're missing the return type (and myabe other ts stuff?)

**Review by alexflorisca — COMMENTED** [review]
Nice work, can't wait for this to develop more and more! Left a few comments inline

File: tests/e2e/specs/performance/cart.test.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		expect( true ).toBe( true );
+	} );
+
+	it( 'Quantity change', async () => {
+		await shopper.block.goToCart();
+		await page.waitForNetworkIdle( { idleTime: 2000 } );
+		await page.waitForSelector(
+			'button.wc-block-components-quantity-selector__button--plus'
+		);
+		let i = 10;
```
Comment (alexflorisca) [inline tests/e2e/specs/performance/cart.test.js]: I'd set this to 3, same as the previous test to 1) speed the test up and 2) to keep consistent data points throughout our tests

File: tests/e2e/specs/performance/cart.test.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+		);
+	} );
+
+	it( 'Coupon entry', async () => {
+		await shopper.block.goToCart();
+		await page.waitForNetworkIdle( { idleTime: 2000 } );
+		await page.waitForSelector(
+			'button.wc-block-components-quantity-selector__button--plus'
+		);
+		let i = 10;
```
Comment (alexflorisca) [inline tests/e2e/specs/performance/cart.test.js]: i = 3, same as above

**Comment by alexflorisca** [comment]
I've implemented all the changes from my feedback. @nielslange if you want to have a look and see if you spot anything and then we can merge :)

File: tests/e2e/config/performance-reporter.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
+/**
+ * External dependencies
+ */
+const { readFileSync, statSync } = require( 'fs' );
+const chalk = require( 'chalk' );
```
Comment (nielslange) [inline tests/e2e/config/performance-reporter.js:5]: Very nice! I love the color formatted entries! 👏

**Review by nielslange — APPROVED** [review]
LGTM! 🚢 

### Additional notes

The third test seems a bit flaky. I run the performance test 12 times on my system and the third test failed twice.

**Pull request merged**